elasticdl/pkg/kernel/kernel_test.go

```diff
@@ -25,3 +26,58 @@ func TestSGD(t *testing.T) {
 
 	assert.Equal(t, b, expected)
 }
+
+func TestAdam(t *testing.T) {
+	const size int = 10
+	grad := make([]float32, size)
+	param := make([]float32, size)
+	m := make([]float32, size)
+	v := make([]float32, size)
+	var lr float64 = 0.1
+	var step int64 = 5
+	var beta1 float64 = 0.9
+	var beta2 float64 = 0.999
+	var epsilon float64 = 1e-8
+
+	for i := 0; i < size; i++ {
+		grad[i] = rand.Float32()
+		param[i] = rand.Float32()
+		m[i] = rand.Float32()
+		v[i] = rand.Float32()
+	}
+
+	expectedParam := make([]float32, size)
+	expectedM := make([]float32, size)
+	expectedV := make([]float32, size)
+
+	for i := 0; i < size; i++ {
+		expectedM[i] = float32(beta1)*m[i] + (1-float32(beta1))*grad[i]
+		expectedV[i] = float32(beta2)*v[i] + (1-float32(beta2))*grad[i]*grad[i]
+	}
+
+	for i := 0; i < size; i++ {
+		expectedParam[i] = param[i] - float32(lr)*expectedM[i]/(1-float32(math.Pow(beta1, float64(step))))/(float32(math.Sqrt(float64(expectedV[i]/(1-float32(math.Pow(beta2, float64(step))))))+epsilon))
+	}
+
+	Adam(grad, param, m, v, lr, int64(size), step, beta1, beta2, epsilon)
+
+	diffParam := 0
+	diffM := 0
+	diffV := 0
+
+	for i := 0; i < size; i++ {
+		if math.Abs(float64(expectedM[i]-m[i])) > 1e-6 {
```

Comment:
It seems that we will compare two float array often, could we add a `Cmp` function to do this?

Here is a reference  https://dev.to/juliaferraioli/testing-in-go-testing-floating-point-numbers-4i0a

Reply:
fixed